hub/src/server/server.ts

```diff
@@ -213,6 +214,23 @@ export class HubServer {
       }
     }
 
+    const requestTag = requestHeaders['if-match']
+    if (!this.driver.supportsETagMatching) {
+      if (requestTag) {
+        const freshTag = (await this.driver.performStat({
```

Comment:
Looking over the associated blockstack.js PR and thinking about the various scenarios during a `putFile` request -- I think we have to include structured data (JSON string) in the `if-match` header in order to handle a couple situations:

Blockstack.js has a client-side {filename,etag} map: 
```ts
const etags: { [filename: string]: string } = {}
``` 

If `putFile` is performed with no corresponding etag, then it could mean at least a couple scenarios:

1. The dev's intention is to create a new file and no existing file expected to exist. However, there may be a concurrency bug _(unanticipated by the dev, as with most concurrency bugs)_ where the new file was already created in another session. I think with the current code, the second "new file" request will have an empty `if-match`, causing this etag check to be skipped and clobbering the original file. 
   * This could be solved if the `if-match` content was structured, for example `JSON.parse(headers['if-match']).etag === null` could indicate that a check should be performed. That check could look something like `(await driver.performStat(...)).exists === false`. 

2. A use-case where file overwrites are intentional, where the local `etag` file map is often outdated -- for example high frequency intentional overwrites coming from multiple sessions. I think we've decided overwrites need to be explicitly specified by devs, with an API that clearly indicates the consequences, e.g. `putFile(..., { dangerousOverwrite: true })`. With this current etag check, blockstack.js would have to implement this by omitting the `if-match` value. 
I think the Gaia http spec should also be explicit about overwrites, and "overwrite by omitting `if-match`" is feels like overloading of legacy behavior. 
   * I think structured `if-match` data could also solve this. For example `JSON.parse(headers['if-match']).overwrite === true` could be checked.

Reply:
In general, I'm in favor of trying to preserve the standard behavior and syntax of the `If-Match` header -- i.e., accepting just an ETAG.

The standard way to deal with the 'new file' case is to use the `If-None-Match` header, as `If-None-Match: *`

Would `blockstack.js` be able to differentiate the intention of creating a new file from the intention of updating a file?

Reply:
`If-None-Match` header is a good idea. I think that could solve the 'new file' scenario. I've generally wanted to avoid new headers due to the additional CORs config overhead, but makes sense in this case. 

> Would blockstack.js be able to differentiate the intention of creating a new file from the intention of updating a file?

At the moment, blockstack.js only knows of the missing local etag map entry, and blockstack.js can only assume the dev's intention is create a new file. This assumption will not always be true. Blockstack.js could implement a `createFile` function (or equivalent option in `putFile`) for devs to specify their intention -- then `putFile` can just throw immediately, rather than assume, attempt upload, and throw. 

---

The remaining problem is how to specify an explicit overwrite. I think this could be performed by specifying `If-Match: *`?

So the possible blockstack.js-vNext `putFile` requests must be one of:
* `If-Match: <etag-value>`: typical file update -- the etag map entry exists.
* `If-None-Match: *` as new file upload -- assumed from missing etag map entry.
* `If-Match: *` as file overwrite -- dev explicitly specified overwrite flag with  `putFile`.